Remove mentions of the "Path Identifier" (#185)
quicwg/multipath: draft-ietf-quic-multipath.md

capitals, as shown here.

We assume that the reader is familiar with the terminology used in
{{QUIC-TRANSPORT}}. In addition, we define the following term:

- Path: refers to the 4-tuple {source IP address, source port number,
  destination IP address, destination port number}. A path refers to
  "network path" used in {{QUIC-TRANSPORT}}. Endpoints identify a path
  by the sequence number of the Destination Connection ID they use over

>   that path.

Author: While the path actually refers to the (perceived) 4-tuple, not sure we say the path only depends on the CID. The CID (sequence number) is used between endpoints to identify a path in PATH_ABANDON/PATH_STATUS frames. However, we can remove this definition if this brings confusion.

the chosen path, as further described in {{setup}}.
In this version of the document, a QUIC server does not initiate the creation
of a path, but it can validate a new path created by a client.

> A new path can only be used once it has been validated. The Destination
> Connection ID is used to associate a packet to a valid path. Further, the

Reviewer: ```suggestion
A new path can only be used once the associated 4-tuple has been validated
by ensuring that the peer is able to receive packets at that address
(see {{Section 8 of RFC9000}}). The Destination Connection ID is used
to associate a packet to a packet number space that is used on a valid path. Further, the
```

> A new path can only be used once it has been validated. Each endpoint
> identifies its paths by the sequence number of the Destination
> Connection ID it uses over them. Such numerical identifier is notably
> used when an endpoint sends a PATH_ABANDON frame to request its peer to
> abandon the path on which the sender uses the Destination Connection ID
> with the sequence number contained in the PATH_ABANDON frame.

Reviewer: ```suggestion
A new path can only be used once it has been validated. The Destination
Connection ID is used to associate a packet to a valid path. Further, the
sequence number of Destination Connection ID is used as numerical identifier
in control frames. E.g. an endpoint sends a PATH_ABANDON frame to request its peer to
abandon the path on which the sender uses the Destination Connection ID
with the sequence number contained in the PATH_ABANDON frame.
```

Reviewer: Maybe we can also move the text about the status frame here.

### Use PATH_ABANDON Frame to Close a Path {#path-abandon-close}

Both endpoints, namely the client and the server, can initiate path closure,

> by sending PATH_ABANDON frame (see {{path-abandon-frame}}) which

Reviewer: ```suggestion
by sending a PATH_ABANDON frame (see {{path-abandon-frame}}) which
```

which requests the peer to stop sending packets with the corresponding Destination Connection ID.
The PATH_ABANDON frame contains the Destination Connection

> ID Sequence Number and therefore can be sent on any path.

Reviewer: This is entirely my fault with PR #174 but this is actually specified already later in the text and therefore redundant here. However, we can also just leave this for now and make another editorial pass later.

Reviewer: Okay, should we open an (editorial) issue, so we don't forget? Or we can create a new PR right after this one is merged (without an issue).

Author: Filled #194.

> Both endpoints, namely the client and the server, can close a path,

Reviewer: ```suggestion
Both endpoints, namely the client and the server, can initiate path closure,
```

by sending PATH_ABANDON frame (see {{path-abandon-frame}}) which

> abandons the path with the corresponding Destination Connection ID.

Reviewer: ```suggestion
which request the peer to stop sending packets with the corresponding Destination Connection ID.
```

destination port) used by the endhost to send packets over the path.

- Associated Destination Connection ID: The Connection ID used to send
packets over the path. The endpoint relies on its sequence number to
maintain a Destination Connection ID-specific packet number space for
sending packets over this path. Packet number considerations as
described in {{Section 12.3 of QUIC-TRANSPORT}} apply within the given

> path.

Reviewer: I don't think we need this additional text.

You don't need the sequence number to identify the packet space; you can also just use the CID directly. You only need the sequence number really in control frames.

Also what consideration in RFC9000 do you want to point to?

Author: Indeed, let's remove that added text.

In Active state, hosts MUST also track the following information:

- Associated Source Connection ID: The Connection ID used to receive

> packets over the path. The endpoint relies on its sequence number to
> receive and acknowledge packets belonging to that Connection ID-specific
> packet number space.

Reviewer: ```suggestion
packets over the path. The endpoint relies on its sequence number to
send oath control information and specifically acknowledge packets belonging to that Connection ID-specific
packet number space.
```

Reviewer 2: Typo in suggested change: oath -> path

Reviewer: fixed

The ACK_MP frame, as specified in {{ack-mp-frame}}, is used to
acknowledge 1-RTT packets.
Compared to the QUIC version 1 ACK frame, the ACK_MP frame additionally
contains the receiver's Destination Connection ID Sequence Number field
to distinguish the Connection ID-specific packet number space.

Acknowledgements of Initial and Handshake packets MUST be carried using
ACK frames, as specified in {{QUIC-TRANSPORT}}. The ACK frames, as defined
in {{QUIC-TRANSPORT}}, do not carry the Destination Connection ID

> Sequence Number field. If multipath has been successfully

Reviewer: ```suggestion
Sequence Number field to identify the packet number space.
If the multipath extension has been successfully
```

The ACK_MP frame, as specified in {{ack-mp-frame}}, is used to
acknowledge 1-RTT packets.
Compared to the QUIC version 1 ACK frame, the ACK_MP frame additionally
contains the Receiver's Destination Connection ID Sequence Number field
to distinguish the Connection ID-specific packet number space
acknowledged.

Acknowledgements of Initial and Handshake packets MUST be carried using
ACK frames, as specified in {{QUIC-TRANSPORT}}. The ACK frames, as defined
in {{QUIC-TRANSPORT}}, do not carry the Receiver's Destination

> Connection ID Sequence Number field. If multipath has been successfully
> negotiated, ACK frames in 1-RTT packets acknowledge packets sent with
> the Connection ID having sequence number 0.

Reviewer: ```suggestion
Connection ID Sequence Number field. If multipath has been successfully
negotiated, ACK frames in 1-RTT packets MUST be interpreted as acknowledging
packets on the path they the packet with the ACK frames been received.
```

Reviewer 2: No, please do not apply this change. This PR is editorial in nature, and should not change the consensus so far, i.e., the old ACK frame acknowledges CID number 0. Quentin's original text is fine.

Reviewer: Yes, this changes goes maybe a bit far for this PR but Quentin's text is not right because he removed the "If for any reason  ACK frames are received in 1-RTT packets while the state of multipath negotiation is ambiguous" part. In this ambiguous state all acks are mapped to the seq number 0. However, this is not true later. My proposal is not to change the consensus but rather say the more general thing.

Author: Yes, let's avoid introducing design changes here. Also, I think relying on the reception 4-tuple may lead to ambiguous state at the receiver.

Author: For the ambiguous part, I'm not sure to understand when the negotiation of multi path could be ambiguous. But so far the less ambiguous approach (as I understood so far) is that ACK frames (always) relate to CID number 0.

Reviewer: I was assuming an ACK frame always corresponds to the path it was sent on. However, it seem this is simply not defined in the draft and we probably need to open an issue to clarify this.

Reviewer 2: No, we specifically do not want implementations to continue sending ACK frames after negotiating multipath. The notion of "current path" is ambiguous -- for example, it does change in case of NAT rebinding. Adding the CID-sequence-number in the ACK_MP guarantee that there is no such ambiguity. As for nake ACK, the equivalence to "cid-number=0" is straightforward.

As soon as the negotiation of multipath support is completed,
endpoints SHOULD use ACK_MP frames instead of ACK frames to acknowledge application
data packets, including 0-RTT packets, using the Connection ID with
sequence number 0 after the handshake concluded.

> ACK_MP frame (defined in {{ack-mp-frame}}) SHOULD be sent on the path
> it perceives the Connection ID of the packet number space it acknowledges.

Reviewer: ```suggestion
ACK_MP frame (defined in {{ack-mp-frame}}) SHOULD be sent on same the path
as the data was sent that it acknowledged.
```

Author: While I understand the suggestion, I'm not sure this makes the text clearer, though I don't have strong opposition here.

Author: I fully agree here. Sending the ACK_MP on the same path is a reasonable default behaviour, but in some network configurations, this is clearly suboptimal and there is no point forcing endpoints to do that. I think having a  "SHOULD" is fine, but not a "MUST".

Rethinking further, I'm wondering if this paragraph could even be removed based on discussions in #183. We already discuss the ACK_MP scheduling in implementation considerations, and that should be enough. Maybe add at some place that frames can effectively be sent and received on any valid/active path.

Reviewer: ```suggestion
ACK_MP frame (defined in {{ack-mp-frame}}) SHOULD be sent on the path
it received packet with the Connection ID of the packet number space it acknowledges.
```

Reviewer: On the editorial part: I think "perceives" is just the wrong word. I was also trying to make it shorter but feel free to take any of the two suggestions.

On the text itself, if we use normative language it should stay here and not move to the implementation guidance section. let keep this for now and maybe open a separate issue in case we need more discussion or actually want to change it.

My view is: I think we need some normative guidance because of the RTT calculation issue. However, there might be room for refining it or provide further explanation in the implementation guidance scheduling section.

Reviewer 2: "receive packet with" is a better text than "perceived".

And yes, let's merge this PR as is, and discuss path affinity of MP_ACK separately.

Reviewer 2: I started a discussion of the "ACK on same path" in issue #190

PATH_STATUS Frames contain the following fields:

Sender's Destination Connection ID Sequence Number:
: The sequence number of the Destination Connection ID used by the
  sender of this frame to send packets over the path the status update

>   corresponds.

Reviewer: ```suggestion
  corresponds to.
```

Compared to the ACK frame specified in {{QUIC-TRANSPORT}}, the following
field is added.

Receiver's Destination Connection ID Sequence Number:
: The sequence number of the Connection ID identifying the packet number
  space of the 1-RTT packets which are acknowledged by the ACK_MP frame.

> 

Reviewer 2: I think we have an issue of consistency. In the Abandon and Path Status frames, the DCID field is called "Sender's DCID". Here, it is called "Receiver's DCID". Yet, they refer to exactly the same value. My suggestion:

* Remove "Sender's" from the DCID descriptions in PATH_ABANDON and PATH_STATUS
* Remove "Receiver's" from the DCID descriptions in ACK_MP
* Otherwise, keep the text as is, because the text is correct.

Reviewer: No they do not refer to the same value. The Sender's CID is the one used when you sent a packet. The Receivers's CID is the one used by the other end to send - hence when you receive packets.

Author: Actually the PATH_ABANDON and PATH_STATUS frame includes the sequence number of the DCID of the sender of the frame, while in ACK_MP the sequence number corresponds to the DCID used by its peer, hence the (verbose) name convention. But if this introduces confusion, I can remove "Sender's" and "Receiver's" mentions.

Reviewer: Actually we did a mistake here when I remove the type field. The path abandon frame should also have the receivers cid as you tell the other end to not use this for sending data anymore. Same I probably true for the path status frame.

Reviewer 2: I think aligning ACK, ABANDON and STATUS makes sense. ABANDON is in many respects just the symmetric of the ACK. For example, a node receives data on a cellular path that charges per byte when a "free" path is available. The node will send the ABANDON immediately, telling its peer to "stop doing that." Using the DCID "used by the peer to send 1-RTT packets" makes a lot of sense.

Author: Yeah, this would make sense. Let me update the text and the examples accordingly.

If an endpoint receives an ACK_MP frame with a Connection ID sequence
number which was never issued (i.e., with a sequence number
larger than the largest one advertised), it MUST treat this as a connection
error of type MP_PROTOCOL_VIOLATION and close the connection.
If an endpoint receives an ACK_MP frame with a Connection ID sequence
number which is no more active (e.g., retired by a RETIRE_CONNECTION_ID
frame or belonging to closed paths), it MUST ignore the ACK_MP frame
without causing a connection error.

> 

Reviewer 2: The reference to "closed paths" is a bit ambiguous, because it is in fact OK to send an ACK_MP after and ABANDON_PATH -- or even in the same packet, but QUIC does not specify the order of processing of frames in a packet. And it is also OK to process it, rather than spuriously repeat data that has been actually received. I would be tempted to just strike out "or belonging to closed paths". Or maybe strike out the whole parenthesis. Or maybe strike out the whole paragraph about not active anymore. For comparison, Section 19.16 of RFC 9000 just says:

Receipt of a RETIRE_CONNECTION_ID frame containing a sequence number greater than any previously sent to the peer MUST be treated as a connection error of type PROTOCOL_VIOLATION.

Author: I think your point about the whole parenthesis is right. Also, this issue about seeing retired DCID also applies in PATH_ABANDON and PATH_STATUS (i.e., just ignore the frame), so we may just want to add a sentence at the beginning of the presentation of the multi path-specific frames?

Reviewer: I see that this text was remove, which is fine. But was it re-added somewhere else?

Author: Not yet. We may add a sentence to clarify that we should just ignore Multipath frames related to retired DCIDs, but I would suggest to add such a sentence at the beginning of the section introducing these frames, as this applies to all of them.

Reviewer: Okay, maybe let's open an issue, so we don't forget!

Author: Filled #193.
